akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala

```diff
@@ -83,24 +86,27 @@ private[http] object HttpServerBluePrint {
   def controller(settings: ServerSettings, log: LoggingAdapter): BidiFlow[HttpResponse, ResponseRenderingContext, RequestOutput, RequestOutput, Unit] =
     BidiFlow.fromGraph(new ControllerStage(settings, log)).reversed
 
-  def requestPreparation(settings: ServerSettings, remoteAddress: Option[InetSocketAddress])(implicit mat: Materializer): BidiFlow[HttpResponse, HttpResponse, RequestOutput, HttpRequest, Unit] = {
+  //  def requestPreparation(settings: ServerSettings, remoteAddress: Option[InetSocketAddress])(implicit mat: Materializer): BidiFlow[HttpResponse, HttpResponse, RequestOutput, HttpRequest, Unit] = {
+  def requestPreparation(settings: ServerSettings)(implicit mat: Materializer): BidiFlow[HttpResponse, HttpResponse, RequestOutput, HttpRequest, Unit] = {
     import settings._
 
     val prepareRequest =
       Flow[RequestOutput]
         .splitWhen(x ⇒ x.isInstanceOf[MessageStart] || x == MessageEnd)
         .via(headAndTailFlow)
+        .exposeAttribute[HttpAttributes.RemoteAddress]
         .map {
-          case (RequestStart(method, uri, protocol, hdrs, createEntity, _, _), entityParts) ⇒
+          case ((RequestStart(method, uri, protocol, hdrs, createEntity, _, _), entityParts), remoteAddress) ⇒
             val effectiveMethod = if (method == HttpMethods.HEAD && transparentHeadRequests) HttpMethods.GET else method
+            val addr = remoteAddress.flatMap(_.address)
             val effectiveHeaders =
-              if (settings.remoteAddressHeader && remoteAddress.isDefined)
-                headers.`Remote-Address`(RemoteAddress(remoteAddress.get)) +: hdrs
+              if (settings.remoteAddressHeader && addr.isDefined)
+                headers.`Remote-Address`(RemoteAddress(addr.get)) +: hdrs
```

Comment:
Since you are using `.get` anyway, how about discarding the flatMap?
~~~scala
if (settings.remoteAddressHeader && remoteAddress.isDefined)
  headers.`Remote-Address`(RemoteAddress(remoteAddress.get.address)) +: hdrs
~~~